### ISSUE TITLE
Joint-Limits Violations

### ISSUE THREAD

**Issue by itaouil** [graiola/wolf-setup#6]
Hi again :),

As started to discuss in #5, there are some joint-limits violations after the new debs update, as shown in the pictures below:

**Starting joint positions (once the solver is on)**

![joints_start](https://user-images.githubusercontent.com/17275225/142918277-18e876ae-1855-4420-976e-4c8bdf13bf62.png)

**After a few steps**:

![joints_stop](https://user-images.githubusercontent.com/17275225/142918374-f0f2f32d-4250-41f0-a7a7-48bd1d38d65b.png)

**Joints warnings received**:

![warnings](https://user-images.githubusercontent.com/17275225/142918548-a920f663-8729-47f9-816c-52826a792753.png)

**Comment by graiola** [comment]
I am going to run some tests in the next days to see what's happening...

**Comment by itaouil** [comment]
Thanks :).

**Comment by graiola** [comment]
I run some tests and it looks like the posture problem it's related to the controller parameters.
Try to use these parameters:

[params.txt](https://github.com/graiola/wbc-setup/files/7589612/params.txt)

Regarding the joint limits, we are investigating that, I let you know asap :)

**Comment by itaouil** [comment]
The new parameters helped for the posture :).

I also took another video in case it helps: https://drive.google.com/file/d/1cP9xFqdr8kj8cB3RrAXXkot6jGqZVueP/view?usp=sharing.

In the video you can see the AlienGo limping. I guess this is still due to the joint violations. The limping also happened before updating the params.

**Comment by EnricoMingo** [comment]
Can you check the velocity references that you are sending? Because it looks like sometimes it takes two steps with the same leg. Maybe you have also to play a little with the parameters of the walking pattern generator.

**Comment by graiola** [comment]
Hi @itaouil , we just updated the code to take into account the joint limits into the wbc, you can use the new debian pkg. It should work better but since the joint limit constraints are working by approximation, it could happen that the controller still throws out some warnings...

Let us know if it improves the behavior of the doggo! :)

**Comment by itaouil** [comment]
Hi,

Sorry for the late reply.

I just tried to test the new debs, but weirdly enough I obtain a similar error as issue #1:

![image](https://user-images.githubusercontent.com/17275225/143462501-43ec11e1-de97-4b9f-9f16-c7ea1cca9e52.png)

When I try the previous commit, the same error does not happen.

**Comment by graiola** [comment]
Maybe I did a mistake while creating the debians, I am going to generate them again in few minutes.

**Comment by graiola** [comment]
@itaouil try again please...

**Comment by itaouil** [comment]
Just pulled down the new debs, and the previous issue is solved.

I will do some tests tomorrow finally, and will update you :).

**Comment by itaouil** [comment]
I just finished testing the the new debs.

I approached it by sending different velocity commands from 0.1 till 0.9 included with a step of 0.1 for 20 seconds for the following motions:

- Forward
- Right step
- Left step
- Clockwise rotation
- Counter clockwise

for both 6 cm and 15 cm step heights.

Except for some warnings of joint violations and step length > 0.4 and two falls (during a continuous in-place rotation at 0.6 and 0.9) in approximately 40+ minutes of testing it worked remarkably well, and the reactive stairs climbing ability is still working great :)

I will be starting collecting some data in simulation for my project so I will further test it over the next days, but I think the wbc has greatly improved in its stability and robustness since I first tried it, except for rare circumstances like continuous in-place rotation at high velocity (which don't always make the robot fall) and which I will try to investigate in the near future.

Thanks again for open-sourcing this amazing work and the help! :)
